### ISSUE TITLE
Provide machine-readable actionable information in deprecation issues

### ISSUE THREAD

**Issue by gwbrown** [elastic/elasticsearch#36324]
While the Deprecation Info API provides human-readable information about the use of deprecated features, it would be nice to allow an external tool, such as Kibana, to semi-automatically make changes in preparation for an upgrade.  Such actions could include reindexing indices, removing deprecated settings, and so on.

Currently, to do this, a tool would have to read the issue description, which is not guaranteed to remain consistent, and would need a table mapping every issue type to an action.  We *may* be able to provide a simpler alternative through the deprecation API.

One option is to add an (optional) `recommended_actions` field, which indicates that an index needs to be reindexed, a setting needs to be removed or changed, etc.  Notably, this puts the onus of recommending actions in Elasticsearch's court.

Another option would be to add a unique identifier to each issue type.  We are unlikely to do this, as it may impact future localization plans (see #35665).  This also does not resolve the problem of an external tool still needing a mapping of issue type to recommended action.

/cc @joshdover

**Comment by joshdover** [comment]
My main concern with only providing a `recommended_action` field is that would potentially slow down the development of new migration features in Kibana's Upgrade Assistant because we would need to wait on Elasticsearch to update responses in order to tie an action in the UI to a specific deprecation warning.

Additionally, this doesn't allow Kibana to make fine-grained decisions about the action. For instance, there could be multiple deprecation types that we initially believe need the same action, but later discover some fundamental difference between them that would change how we process that action. In that case, Kibana would need to fallback again to reading the human-readable (and unstable) `message` field to determine how to process the action. Related is the concern that these recommended actions could be too generic and re-used in ways that are not specific enough.

All this said, I don't anticipate adding many actions in Kibana at this time. But I also don't think this actually overlaps with the concerns in #35665, which primarily concerns exception messages. Exceptions are prevalent in the Elasticsearch codebase and I agree that changing all of those to add an error code field would be a large undertaking. However, the Deprecations Info API does not display errors but instead deprecation messages for future versions of Elasticsearch. Introducing a pattern here, whether its a `recommended_action` field or a `deprecation_code` field, I don't think it introduces an inconsistency from the rest of the API surface.

I argue that adding a unique `deprecation_code` is more flexible for downstream consumers, allowing Kibana to iterate faster on new migration features without relying on brittle human-readable strings.

I certainly don't want to over-engineer this, and more than anything this is a nice-to-have from a maintainability standpoint. For example, in the API's current form, I will have to have a different string to check between the 6.x branches and 7.x branches ("Index was created before X.X"), but this isn't the end of the world. I'm just concerned about forgetting about it and the feature breaking a year or more from now.

**Comment by joshdover** [comment]
For the time being, we're going to rely on string matches and will include functional tests to ensure these do not break in future releases. Going to close this.
